lib/utility/error_monitor.rb

```diff
@@ -92,10 +91,32 @@ def raise_if_necessary
     end
 
     def num_errors_in_window
-      @window_errors.count(&:itself).to_f
+      @window_errors.count(true).to_f
     end
 
-    def increment_window_index
+    def track_window_error(is_error)
+      # We keep the errors array of the size @window_size this way, imagine @window_size = 5
+      # Error array inits as falses:
+      # [ false, false, false, false, false ]
+      # Third document raises an error:
+      # [ false, false, true,  false, false ]
+      #                 ^^^^
+      #                 2 % 5 == 2
+      # Fifth document raises an error:
+      # [ false, false, true,  false, true  ]
+      #                               ^^^^
+      #                               4 % 5 == 4
+      # Sixth document raises an error:
+      # [ true, false, true,   false, true  ]
+      #   ^^^^
+      #   5 % 5 == 0
+      #
+      # Eigth document is successful:
+      # [ true, false, false,  false, true  ]
+      #                ^^^^^
+      #                7 % 5 == 2
+      # And so on.
+      @window_errors[@window_index] = is_error
```

Comment:
I will be more defensive here, but not a blocker.

Reply:
I copied the code from Enterprise Search and tried to leave it as close to original as possible, but in the end I'd rewrite this class fully - it's a bit sloppy and as research showed it's prone to bugs.

Reply:
oof, that's a little hurtful. I was quite proud of this implementation. Of course, didn't catch this bug of overwriting past errors in the window to successes 🙈 . Nice fix, and great comment to explain the logic.

Reply:
Sorry @seanstory, when writing my comment I did not think that it could be hurtful, but I see it now.

I had some small problems when reading the code when the bug happened and figuring out the logic and it made me think that some parts of the code can be encapsulated better - particularly the rate calculating logic - but the code works in every known scenario now, so it's a working component that we can use. Can it be improved? Of course, any piece of code can :)